src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+                batch_tmp_path.display(),
+                batch_path.display()
+            )
+        })?;
+
+        Ok(Batch {
+            upper,
+            lower,
+            path: batch_path,
+        })
+    }
+
+    /// Reintroduce a batch based on an available file in the trace
+    /// directory.
+    fn reinit(path: PathBuf) -> Result<Self, anyhow::Error> {
+        let batch_name = path
+            .file_name()
+            .expect("batch name known to exist")
+            .to_str()
+            .expect("batch name known to be valid utf8");
+        let parts: Vec<_> = batch_name.split('-').collect();
+        assert!(parts.len() == 3);
+        Ok(Self {
+            upper: parts[2].parse()?,
+            lower: parts[1].parse()?,
+            path,
+        })
+    }
+
+    /// Read the data from a batch stored on disk into memory.
+    fn read(&self) -> Result<Vec<Message>, anyhow::Error> {
+        read_segment(&self.path)
+    }
+
+    /// Physically concatenate all batches together into a single batch
+    /// and consolidate updates up to the frontier if provided.
+    fn compact(
+        batches: &[Batch],
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
+    ) -> Result<Self, anyhow::Error> {
+        let mut messages = vec![];
+
+        for batch in batches {
+            messages.append(&mut read_segment(&batch.path)?);
+        }
+
+        Batch::create_from_messages(messages, trace_path, compaction_frontier)
+    }
+}
+
+/// A Trace is an on-disk representation of data meant to mimic a differential Trace.
+///
+/// A Trace listens to the `wal_path` and checks for WAL segments that it can consolidate
+/// into Batches (stored in the `trace_path`). Once it exceeds a certain number of Batches
+/// it tries to physically and logically compact them into a single batch that is
+/// compacted up to the `compaction` frontier.
+#[derive(Debug)]
+pub struct Trace {
+    trace_path: PathBuf,
+    wal_path: PathBuf,
+    batches: Vec<Batch>,
+    compaction: Option<Timestamp>,
+}
+
+impl Trace {
+    fn create(id: GlobalId, trace_path: PathBuf, wal_path: PathBuf) -> Result<Self, anyhow::Error> {
+        let _ = fs::read_dir(&wal_path).with_context(|| {
+            format!(
+                "trying to ensure wal directory {} exists for trace of relation {}",
+                id,
+                wal_path.display()
+            )
+        })?;
+
+        // Create a new directory to store the trace
+        fs::create_dir(&trace_path).with_context(|| {
+            format!("trying to create trace directory: {}", trace_path.display())
+        })?;
+
+        Ok(Self {
+            trace_path,
+            wal_path,
+            batches: Vec::new(),
+            compaction: None,
+        })
+    }
+
+    /// Remove all on-disk data for this trace.
+    ///
+    /// Importantly, we also delete the WAL directory here (the WAL writer
+    /// only gets to add new files and can't do anything else)
```

Comment:
Brutal. Is this important to do here, or just happens that it might as well?

Reply:
added documentation. i think this was the simplest design i could think of without having having the compacter relay information about delete completions back to the coordinator / wal writer